src/coreclr/jit/emit.h

```diff
@@ -2288,31 +2292,134 @@ class emitter
+                                                                             // new IG, and it's an extension IG.
+                    ((prevIG->igFlags & IGF_NOGCINTERRUPT) == (ig->igFlags & IGF_NOGCINTERRUPT)));
+        }
+        else if (prevIG != ig)
+        {
+            return ((prevIG->igFlags & IGF_EXTEND) &&
+                    ((prevIG->igFlags & IGF_NOGCINTERRUPT) == (ig->igFlags & IGF_NOGCINTERRUPT)));
+        }
+        return true;
+    }
+
     // Check if a peephole optimization involving emitLastIns is safe.
     //
     // We have the following checks:
     // 1. There must be a non-null emitLastIns to consult (thus, we have a known "last" instruction).
     // 2. `emitForceNewIG` is not set: this prevents peepholes from crossing nogc boundaries where
     //    the next instruction is forced to create a new IG.
-    // 3. Looking backwards across an IG boundary can only be done if we're in an extension IG.
-    // 4. The IG of the previous instruction must have the same GC interrupt status as the current IG.
-    //    This is related to #2; it disallows peephole when the previous IG is GC and the current is NOGC.
-    bool emitCanPeepholeLastIns()
+    bool emitCanPeepholeLastIns() const
     {
-        assert((emitLastIns == nullptr) == (emitLastInsIG == nullptr));
+        assert(emitHasLastIns() == (emitLastInsIG != nullptr));
+
+        return emitHasLastIns() && // there is an emitLastInstr
+               !emitForceNewIG &&  // and we're not about to start a new IG.
+               canPeepholeCrossIGBoundaries(emitLastInsIG, emitCurIG);
+    }
 
-        return (emitLastIns != nullptr) && !emitForceNewIG &&
-               ((emitCurIGinsCnt > 0) ||                     // we're not at the start of a new IG
-                ((emitCurIG->igFlags & IGF_EXTEND) != 0)) && //    or we are at the start of a new IG,
-                                                             //    and it's an extension IG
-               ((emitLastInsIG->igFlags & IGF_NOGCINTERRUPT) == (emitCurIG->igFlags & IGF_NOGCINTERRUPT));
+    enum emitPeepholeResult
+    {
+        PEEPHOLE_CONTINUE,
+        PEEPHOLE_ABORT
+    };
+
+    // Visits the last emitted instructions.
+    // Must be safe to do - use emitCanPeepholeLastIns for checking.
+    // Action is a function type: instrDesc* -> emitPeepholeResult
+    template <typename Action>
+    void emitPeepholeIterateLastInstrs(Action action)
+    {
+        assert(emitCanPeepholeLastIns());
+
+#if EMIT_BACKWARDS_NAVIGATION
+        insGroup*  ig;
+        instrDesc* id;
+
+        if (!emitGetLastIns(&ig, &id))
+            return;
+
+        for (unsigned i = 0; i < EMIT_MAX_PEEPHOLE_INS_COUNT; i++)
+        {
+            assert(id != nullptr);
+
+#if !defined(TARGET_LOONGARCH64)
+            switch ((ID_OPS)emitFmtToOps[id->idInsFmt()])
+            {
+                // We cannot safely look at previous instructions at these boundaries.
```

Comment:
This is being pessimistic, but it's definitely *true* for determining if the upper 32 bits are zero.

Would you rather have this kind of check in `AreUpper32BitsZero`?

Reply:
Can you tell me _why_ "we cannot safely look at previous instructions at these boundaries"? Even for "AreUpper32BitsZero"?

Reply:
We can't determine if the register's upper 32-bits are zero when the previous instruction is a branching operation because we cannot determine the control flow of the register at this point, so we have to bail.

Reply:
I guess I don't understand that. If you have:
```
movzx rax, eax (1)
<conditional branch> L_123 // branch somewhere
movzx rax, eax // (2)
```
then if control reaches (2), the upper 32 bits of rax are guaranteed to be zero. The branch doesn't affect them.

What does matter is if there is a branch *target* between (1) and (2). But that is handled by not allowing peeps to cross non-IGF_EXTEND IG boundaries.

Reply:
It seems plausible, if we know there isn't a label in-between:
```
movzx rax, eax (1)
<conditional branch> L_123 // branch somewhere
<label>
movzx rax, eax // (2)
```
^ - I wouldn't deem this safe to skip.